File: hardToTest/PerfectNumber/PerfectNumber.py
import math

def is_prime(p):
    for i in range(2, int(math.sqrt(p)) + 1):
        if p % i == 0:
            return False
    return p > 1  # p must be greater than 1

def is_perfect_number(num):
    # Perfect numbers must be positive integers
    if num <= 1:
        return False

    # Check if num is even
    if num % 2 == 1:
        return False

    # Find p such that num = 2^(p-1) * (2^p - 1)
    for p in range(2, 32):  # p needs to be less than 32 to avoid overflow
        if is_prime(p) and (2 ** p - 1) % 2 == 1:
            perfect_candidate = (2 ** (p - 1)) * (2 ** p - 1)
            if perfect_candidate == num:
                return True

    return False

if __name__ == "__main__":
    num = 28
    if is_perfect_number(num):
        print(f"{num} 是完全数。")
    else:
        print(f"{num} 不是完全数。")

File: hardToTest/PerfectNumber/PerfectNumber.c
#include <stdio.h>
#include <math.h>
#include <stdbool.h>

bool isPrime(int p) {
    for (int i = 2; i <= sqrt(p); i++) {
        if (p % i == 0) {
            return false;
        }
    }
    return p > 1; // p must be greater than 1
}

bool isPerfectNumber(int num) {
    // Perfect numbers must be positive integers
    if (num <= 1) {
        return false;
    }

    // Check if num is even
    if (num % 2 == 1) {
        return false;
    }

    // Find p such that num = 2^(p-1) * (2^p - 1)
    for (int p = 2; p < 32; p++) { // p needs to be less than 32 to avoid overflow
        if (isPrime(p) && (pow(2, p) - 1) != 0) {
            int perfectCandidate = (int)(pow(2, p - 1) * (pow(2, p) - 1));
            if (perfectCandidate == num) {
                return true;
            }
        }
    }

    return false;
}

int main() {
    int num = 28;
    if (isPerfectNumber(num)) {
        printf("%d 是完全数。\n", num);
    } else {
        printf("%d 不是完全数。\n", num);
    }
    return 0;
}